File: python/crypt_arithm.py
from itertools import permutations

def letter_maps(letters):
    # Generate all possible mappings from letters to digits 
    perms =  permutations('0123456789', len(letters))
    return (dict(zip(letters, digits)) for digits in perms)

def replace_letters(crypt, letter_map):
    return ''.join(letter_map.get(c, c) for c in crypt)

def crypt_letters(crypt):
    return set(c for c in crypt if c.isalpha())

def candidates(crypt):
    maps = letter_maps(crypt_letters(crypt))
    return (replace_letters(crypt, m) for m in maps)

def is_solution(candidate):
    try:
        return eval(candidate.replace('=', '=='))
    except:
        return False
    
def solve(crypt):
    """ Return all possible solutions to the crypt in iterator

    solve is a generator that lazily generates all possible solutions 
    to the given crypt
    """
    return (c for c in candidates(crypt) if is_solution(c))

def main():
    crypt = "SEND + MORE = MONEY"
    print "Solving: ", crypt
    print "Letters: ", ', '.join(sorted(crypt_letters(crypt)))

    print "\nPossible solutions"
    for s in solve(crypt):
        print s

if __name__ == "__main__":
    main()

File: haskell/CryptArithm.hs
import Data.List hiding (permutations)
import qualified Data.Map as Map 
import Data.Char

permutations :: Eq a => [a] -> Int -> [[a]]
permutations xs len 
    | len < 0         = error("permutations: len cannot be negative")
    | length xs < len = []
    | len == 0        = [[]]
    | otherwise       = [x:y | x <- xs, y <- permutations (delete x xs) (len - 1)]

letterMaps :: [Char] -> [Map.Map Char Char]
letterMaps letters = [Map.fromList $ zip letters digits | digits <- perms]
    where perms = permutations ['0'..'9'] (length letters)

replace :: [Char] -> Map.Map Char Char -> [Char]
replace letters letterMap = map mapLetter letters 
    where mapLetter l = Map.findWithDefault l l letterMap

candidates :: String -> [String]
candidates crypt = [replace crypt letterMap | letterMap <- letterMaps letters]
    where letters = nub $ filter isAlpha crypt

isSolution :: String -> Bool
isSolution = check . words
    where check [a, "+", b, "=", c] 
              | "0" `isPrefixOf` a  = False 
              | "0" `isPrefixOf` b  = False 
              | "0" `isPrefixOf` c  = False 
              | otherwise         = (read a) + (read b) == (read c)

solve :: String -> [String]
solve = (filter isSolution) . candidates

crypt = "SEND + MORE = MONEY"

main = do
    putStrLn $ "Solving: " ++ crypt
    let solutions = solve crypt
    putStrLn $ "Found " ++ show (length solutions) ++ " solutions"
    mapM_ putStrLn solutions
 
